targets/csharp/source/source/Uunit/PluginManagerTest.cs

```diff
@@ -0,0 +1,353 @@
+            var customTransport = new CustomTransportPlugin();
+
+            try
+            {
+                // Set a custom transport plugin
+                PluginManager.SetPlugin(customTransport, PluginContract.PlayFab_Transport);
+
+                // Call some PlayFab API 
+                var getRequest = new GetUserDataRequest();
+                var getDataTask = PlayFabClientAPI.GetUserDataAsync(getRequest, null, extraHeaders);
+                ContinueWithContext(getDataTask, testContext, null, false, "GetUserData call failed", false);
+                getDataTask.Wait();
+
+                // Verify
+                testContext.True(customTransport.IsDoPostCalled);
+                testContext.EndTest(UUnitFinishState.PASSED, null);
+            }
+            catch (Exception e)
+            {
+                testContext.EndTest(UUnitFinishState.FAILED, e.ToString());
+            }
+            finally
+            {
+                // Restore the original plugin
+                PluginManager.SetPlugin(playFabTransport, PluginContract.PlayFab_Transport);
+            }
+        }
+
+        /// <summary>
+        /// CLIENT AND SERVER API
+        /// Test that multiple plugins can be used with the same contract simultaneously.
+        /// </summary>
+        [UUnitTest]
+        public void SdkMultiplePluginsPerContract(UUnitTestContext testContext)
+        {
+            // Create multiple custom plugins, set them with the same contract
+            const int pluginNumber = 4;
+            var APIs = new CustomApi[pluginNumber];
+            for (int i = 0; i < pluginNumber; i++)
+            {
+                var pluginName = "Plugin_" + i;
+                var customTransport = new CustomTransportPlugin() { Name = pluginName };
+                PluginManager.SetPlugin(customTransport, PluginContract.PlayFab_Transport, pluginName);
+                APIs[i] = new CustomApi() { PluginName = pluginName };
+            }
+
+            try
+            {
+                // Call the APIs
+                int totalNumberOfCalls = 0;
+                var resultTasks = new ConcurrentBag<Task<object>>();
+                Parallel.ForEach(APIs, api =>
+                {
+                    // Each API uses its specific transport plugin
+                    Task<object> task = api.SomeApiMethod();
+                    resultTasks.Add(task);
+                    totalNumberOfCalls++;
+                });
+
+                var resultTasksArray = resultTasks.ToArray();
+                Task.WaitAll(resultTasksArray);
+
+                // Verify
+                testContext.True(totalNumberOfCalls == pluginNumber);
+                for (int i = 0; i < pluginNumber; i++)
+                {
+                    var pluginName = "Plugin_" + i;
+                    var transportPlugin = PluginManager.GetPlugin(PluginContract.PlayFab_Transport, pluginName);
```

Comment:
Right here is a great example. I feel like GetPlugin should be a generic function, so that you don't have to cast it yourself in the next line.
PluginManager should do all the otherwise copy-paste work, so that the caller doesn't have to repeatedly do anything like recasting.